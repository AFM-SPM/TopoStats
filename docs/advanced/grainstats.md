# Grainstats

## At a Glance - Measures Objects

TopoStats automatically tries to measure the grains (objects of interest) found in the grain finding section, in your
AFM images, and outputs them into the `all_statistics.csv` file.

The metrics are briefly summarised in the table below:

| Column Name                      | Description                                                                                                                                                | Data Type |
| -------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- | --------- |
| `center_x/y`                     | The center of the grain.                                                                                                                                   | `float`   |
| `radius_min/max/mean/median`     | The distance from the center to each pixel on the perimeter.                                                                                               | `float`   |
| `height_min/max/mean/median`     | The pixel values underlying the grain mask.                                                                                                                | `float`   |
| `area`                           | The area of the pixel-wise grain mask.                                                                                                                     | `float`   |
| `volume`                         | Volume of the pixel-wise grain mask.                                                                                                                       | `float`   |
| `area_cartesian_bbox`            | The area of a box bounding the grain along cardinal directions.                                                                                            | `float`   |
| `smallest_bounding_width/length` | The shortest bounding box length and perpendicular width of the grain in non-cardinal directions.                                                          | `float`   |
| `smallest_bounding_area`         | The area of the smallest possible box bounding the grain.                                                                                                  | `float`   |
| `aspect_ratio`                   | Ratio of the smallest bounding width to smallest bounding length.                                                                                          | `float`   |
| `max/min_feret`                  | The largest and shortest distance of the calipers rotating the grain between calipers. See [feret diameter](https://en.wikipedia.org/wiki/Feret_diameter). | `float`   |

&nbsp;

![Grain Stats image table pt1](../_static/images/grainstats/ts2_gs_metrics.png)
