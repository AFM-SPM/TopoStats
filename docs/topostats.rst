topostats package
=================

Subpackages
-----------

.. toctree::
   :maxdepth: 4

   topostats.logs
   topostats.tracing

Submodules
----------

.. toctree::
   :maxdepth: 4

   topostats.entry_point
   topostats.filters
   topostats.grains
   topostats.grainstats
   topostats.io
   topostats.plotting
   topostats.plottingfuncs
   topostats.processing
   topostats.run_topostats
   topostats.scars
   topostats.statistics
   topostats.theme
   topostats.thresholds
   topostats.utils
   topostats.validation

Module contents
---------------

.. automodule:: topostats
   :members:
   :undoc-members:
   :show-inheritance:
