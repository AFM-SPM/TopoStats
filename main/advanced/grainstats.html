
<!DOCTYPE html>


<html lang="en" data-content_root="../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Grainstats &#8212; TopoStats 2.3.2.dev596+g13ec2e9a2 documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  <!--
    this give us a css class that will be invisible only if js is disabled
  -->
  <noscript>
    <style>
      .pst-js-only { display: none !important; }

    </style>
  </noscript>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=8878045cc6db502f8baf" rel="stylesheet" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=8f2a1f02" />
    <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=4ae1632d" />
    <link rel="stylesheet" type="text/css" href="../_static/custom.css?v=2b4ec01c" />
  
  <!-- So that users can add custom icons -->
  <script src="../_static/scripts/fontawesome.js?digest=8878045cc6db502f8baf"></script>
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf" />

    <script src="../_static/jquery.js?v=5d32c60e"></script>
    <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
    <script src="../_static/documentation_options.js?v=a6dcfaca"></script>
    <script src="../_static/doctools.js?v=9bcbadda"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'advanced/grainstats';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  <meta name="docsearch:version" content="2.3" />
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <dialog id="pst-search-dialog">
    
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form>
  </dialog>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
<div class="bd-header__inner bd-page-width">
  <button class="pst-navbar-icon sidebar-toggle primary-toggle" aria-label="Site navigation">
    <span class="fa-solid fa-bars"></span>
  </button>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  
    
  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">TopoStats 2.3.2.dev596+g13ec2e9a2 documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../introduction.html">
    Introduction
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../usage.html">
    Usage
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../configuration.html">
    Configuration
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>

            <li class="nav-item dropdown">
                <button class="btn dropdown-toggle nav-item" type="button"
                data-bs-toggle="dropdown" aria-expanded="false"
                aria-controls="pst-nav-more-links">
                    More
                </button>
                <ul id="pst-nav-more-links" class="dropdown-menu">
                    
<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../notebooks.html">
    Notebooks
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../advanced.html">
    Advanced Documentation
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../workflow.html">
    Workflow
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../data_dictionary.html">
    Data Dictionary
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../related_software.html">
    Related Software
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../glossary.html">
    Glossary
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class=" ">
  <a class="nav-link dropdown-item nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

                </ul>
            </li>
            
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
        </div>
      
      
        <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

<button class="btn search-button-field search-button__button pst-js-only" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
 <i class="fa-solid fa-magnifying-glass"></i>
 <span class="search-button__default-text">Search</span>
 <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
</button>
    </div>
  

  
    <button class="pst-navbar-icon sidebar-toggle secondary-toggle" aria-label="On this page">
      <span class="fa-solid fa-outdent"></span>
    </button>
  
</div>

    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
        
      
      <dialog id="pst-primary-sidebar-modal"></dialog>
      <div id="pst-primary-sidebar" class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          
          
            <div class="navbar-item">
<nav>
  <ul class="bd-navbar-elements navbar-nav">
    
<li class="nav-item ">
  <a class="nav-link nav-internal" href="../introduction.html">
    Introduction
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../installation.html">
    Installation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../usage.html">
    Usage
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../configuration.html">
    Configuration
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../contributing.html">
    Contributing
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../notebooks.html">
    Notebooks
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../advanced.html">
    Advanced Documentation
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../workflow.html">
    Workflow
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../data_dictionary.html">
    Data Dictionary
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../related_software.html">
    Related Software
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../glossary.html">
    Glossary
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../faq.html">
    Frequently Asked Questions
  </a>
</li>


<li class="nav-item ">
  <a class="nav-link nav-internal" href="../autoapi/index.html">
    API Reference
  </a>
</li>

  </ul>
</nav></div>
          
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button pst-js-only" aria-label="Color mode" data-bs-title="Color mode"  data-bs-placement="bottom" data-bs-toggle="tooltip">
  <i class="theme-switch fa-solid fa-sun                fa-lg" data-mode="light" title="Light"></i>
  <i class="theme-switch fa-solid fa-moon               fa-lg" data-mode="dark"  title="Dark"></i>
  <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"  title="System Settings"></i>
</button></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<h3>Versions</h3>
<ul>
  <li><a href="../../v2.0.0/index.html">v2.0.0</a></li>
  <li><a href="../../v2.1.0/index.html">v2.1.0</a></li>
  <li><a href="../../v2.1.1/index.html">v2.1.1</a></li>
  <li><a href="../../v2.1.2/index.html">v2.1.2</a></li>
  <li><a href="../../v2.2.0/index.html">v2.2.0</a></li>
  <li><a href="../../v2.2.1/index.html">v2.2.1</a></li>
  <li><a href="../../v2.2.post0/index.html">v2.2.post0</a></li>
  <li><a href="../../v2.3.0/index.html">v2.3.0</a></li>
  <li><a href="../../v2.3.1/index.html">v2.3.1</a></li>
  <li><a href="grainstats.html">main</a></li>
</ul>
</div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
      <div class="sidebar-primary-item">
<div id="ethical-ad-placement"
      class="flat"
      data-ea-publisher="readthedocs"
      data-ea-type="readthedocs-sidebar"
      data-ea-manual="true">
</div></div>
  </div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">

<nav aria-label="Breadcrumb" class="d-print-none">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    <li class="breadcrumb-item active" aria-current="page"><span class="ellipsis">Grainstats</span></li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section id="grainstats">
<h1>Grainstats<a class="headerlink" href="#grainstats" title="Link to this heading">#</a></h1>
<section id="at-a-glance-measures-objects">
<h2>At a Glance - Measures Objects<a class="headerlink" href="#at-a-glance-measures-objects" title="Link to this heading">#</a></h2>
<p>TopoStats automatically measures the grains (objects of interest) found in the grain finding section, in your
AFM images, and outputs them into the <code class="docutils literal notranslate"><span class="pre">grain_statistics.csv</span></code> file.</p>
<p>The metrics are briefly summarised in the table below:</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Column Name</th>
<th>Description</th>
<th>Data Type</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>center_x/y</code></td>
<td>The center of the grain.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>radius_min/max/mean/median</code></td>
<td>The distance from the center to each pixel on the perimeter.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>height_min/max/mean/median</code></td>
<td>The pixel values underlying the grain mask.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>area</code></td>
<td>The area of the pixel-wise grain mask.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>volume</code></td>
<td>Volume of the pixel-wise grain mask.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>area_cartesian_bbox</code></td>
<td>The area of a box bounding the grain along cardinal directions.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>smallest_bounding_width/length</code></td>
<td>The shortest bounding box length and perpendicular width of the grain in non-cardinal directions.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>smallest_bounding_area</code></td>
<td>The area of the smallest possible box bounding the grain.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>aspect_ratio</code></td>
<td>Ratio of the smallest bounding width to smallest bounding length.</td>
<td><code>float</code></td>
</tr>
<tr>
<td><code>max/min_feret</code></td>
<td>The largest and shortest distance of the calipers rotating the grain between calipers. See <a href="https://en.wikipedia.org/wiki/Feret_diameter">feret diameter</a>.</td>
<td><code>float</code></td>
</tr>
</tbody>
</table>
<p> </p>
<p><img alt="Grain Stats image table pt1" src="../_images/ts2_gs_metrics.png" /></p>
</section>
<section id="the-grain-statistics-csv-file">
<h2>The <code class="docutils literal notranslate"><span class="pre">grain_statistics.csv</span></code> file<a class="headerlink" href="#the-grain-statistics-csv-file" title="Link to this heading">#</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">grain_statistics.csv</span></code> file lists metrics for each unique combination of <code class="docutils literal notranslate"><span class="pre">grain_number</span></code>, <code class="docutils literal notranslate"><span class="pre">class_number</span></code>,
and <code class="docutils literal notranslate"><span class="pre">subgrain_number</span></code>, described below. Note that <code class="docutils literal notranslate"><span class="pre">class_number</span></code> will always be constant if standard grain finding
or a binary U-Net model is used for segmentation, but more classes will be included if
multi-class segmentation approaches are used. You can define custom class names within the <code class="docutils literal notranslate"><span class="pre">grainstats</span></code> section
of the config file to replace numeric values in the <code class="docutils literal notranslate"><span class="pre">class_number</span></code> column with meaningful labels.
This can be particularly useful for downstream analysis of extracted metrics.</p>
<table border="1" class="docutils">
<thead>
<tr>
<th>Column Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>grain_number</code></td>
<td>Index of the original grain identified through standard grain finding.</td>
</tr>
<tr>
<td><code>class_number</code></td>
<td>Class assigned during multi-class segmentation.</td>
</tr>
<tr>
<td><code>subgrain_number</code></td>
<td>Index of a sub-region (subgrain) within a grain for a specific class. <br> If multiple regions of the same class are found in one grain, they are numbered sequentially.</td>
</tr>
</tbody>
</table>
<p>A visual example of the labelling procedure is provided below and is
described in greater depth within <code class="docutils literal notranslate"><span class="pre">grain_finding</span></code> documentation.
<img alt="grain-class-subgrain" src="../_images/methods-09.png" /></p>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
</div>
                </footer>
              
            </div>
            
            
              
                <dialog id="pst-secondary-sidebar-modal"></dialog>
                <div id="pst-secondary-sidebar" class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
<div
    id="pst-page-navigation-heading-2"
    class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc" aria-labelledby="pst-page-navigation-heading-2">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#at-a-glance-measures-objects">At a Glance - Measures Objects</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#the-grain-statistics-csv-file">The <code class="docutils literal notranslate"><span class="pre">grain_statistics.csv</span></code> file</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/advanced/grainstats.md.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script defer src="../_static/scripts/bootstrap.js?digest=8878045cc6db502f8baf"></script>
<script defer src="../_static/scripts/pydata-sphinx-theme.js?digest=8878045cc6db502f8baf"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, TopoStats authors.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 8.2.3.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  <!-- # L10n: Setting the PST URL as an argument as this does not need to be localized -->
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.16.1.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>