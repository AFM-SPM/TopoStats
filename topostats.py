"""
File for tracking hard-coded variables (for now)
"""

import argparse

### List of hard-coded variables in pywgwytracing.py

# global variables
gwy_colobar_ln48 = "/module/pixmap/ztype"
gwy_linematchmethod_ln51= '/module/linematch/method'
gwy_linematchmaxdegree_ln52 = "/module/linematch/max_degree"
gwy_polylevel_coldegree_ln53 = "/module/polylevel/col_degree"

# heightediting()
datafield_grains_mark_height_ln128 = int()
gwy_polylevel_masking_ln131 = int()
gwy_linematch_masking_ln135 = int()
filterwidth_ln144 = float()

# editfile()
datafield_markheight_ln170 = int()
gwy_polylevel_masking_ln173 = int()
gwy_linematch_masking_ln177 = gwy_linematch_masking_ln135
gwy_polylevel_masking_ln181 = int()
datafield_gaussfilter_ln195 = int()

# grainfinding()
grain_gauss_ln211 = float()
datafield_mask_outliers_ln220 = float()

# remove_large_objects()
datafield_mask_outliers_ln257 = float()

# remove_small_objects()
datafield_mask_outliers_ln283 = float()

# grain_thinning()
gaussian_size_ln488 = float()

# savefiles()
savefile_type_ln535 = ["/module/pixmap/xytype"]

# saveunknownfiles()
savefile_type_ln573 = ["/module/pixmap/xytype"]

# savecroppedfiles()
savefile_type_ln645 = ["/module/pixmap/xytype"]

# __main__()
path_ln709 = str()
fileend_ln712 = []
filetype_ln713 = str()
extension_ln715 = str()
minheightscale_ln717 = float()
maxheightscale_ln718 = float()
minarea_ln722 = float()
maxdeviation_ln726 = float()
mindeviation_ln727 = float()
cropwidth_ln733 = float()
splitwidth_ln734 = float()
bins_ln736 = int()
channels_ln755 = list()


### dnatracing.py

# getNumpyArraysfromGwyddion()
sigma_ln94 = float()

# getDisorderedTrace()
sigma_ln109 = float()

# purgeObviousCrap()
maxlen_ln126 = int()

# getFittedTraces()
index_width_ln212 = int()

# getSplinedTraces()
step_size_ln376 = float()
interp_size_ln377 = float()

### tracingfuncs.py

# I didn't find any hard-coded variables so far

### statsplotting.py
### TFOplot.py
### traceplotting.py

# not checking these for now due to ongoing merging project
