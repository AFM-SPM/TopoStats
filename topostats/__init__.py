"""Topostats"""
from importlib.metadata import version

from .logs.logs import setup_logger

LOGGER = setup_logger()

release = version("topostats")
__version__ = ".".join(release.split("."[:2]))
