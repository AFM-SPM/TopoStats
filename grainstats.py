import marimo

__generated_with = "0.14.17"
app = marimo.App(width="medium")


@app.cell
def _():
    return


@app.cell
def _():
    import marimo as mo
    return (mo,)


@app.cell(hide_code=True)
def _(mo):
    mo.md(
        r"""
    #load
    #filters
    #grains
    #stats
    """
    )
    return


@app.cell
def _():
    from pathlib import Path
    from topostats.io import LoadScans, find_files, read_yaml
    return Path, find_files


@app.cell
def _(Path, find_files):
    # Set the base directory to be current working directory of the Notebook
    BASE_DIR = Path().cwd()
    print(BASE_DIR)
    # Alternatively if you know where your files area comment the above line and uncomment the below adjust it for your use.
    # BASE_DIR = Path("/path/to/where/my/files/are")
    # Adjust the file extension appropriately.
    FILE_EXT = ".spm"
    # Search for *.spm files one directory level up from the current notebooks
    image_files = find_files(base_dir=BASE_DIR.parent / "tests", file_ext=FILE_EXT)
    return (image_files,)


@app.cell
def _(image_files):
    print(image_files)
    return


if __name__ == "__main__":
    app.run()
