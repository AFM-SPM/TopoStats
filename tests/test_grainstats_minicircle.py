"""Tests for the grainstats module."""
from pathlib import Path

import numpy as np

from topostats.grainstats import GrainStats

BASE_DIR = Path.cwd()
RESOURCES = BASE_DIR / "tests" / "resources"


def test_grainstats_regression(regtest, minicircle_grainstats: GrainStats) -> None:
    """Regression tests for grainstats."""
    statistics, _ = minicircle_grainstats.calculate_stats()
    print(statistics.to_string(), file=regtest)


# @pytest.mark.mpl_image_compare(baseline_dir="resources/img/")
def test_cropped_image(minicircle_grainstats: GrainStats):
    """Tests that produced cropped images have not changed."""
    grain_centre = 547, 794  # centre of grain 7
    length = int(minicircle_grainstats.cropped_size / (2 * minicircle_grainstats.pixel_to_nanometre_scaling))
    cropped_grain_image = minicircle_grainstats.get_cropped_region(
        image=minicircle_grainstats.data, length=length, centre=np.asarray(grain_centre)
    )
    assert cropped_grain_image.shape == (81, 81)

    expected = np.load(RESOURCES / "test_cropped_grains.npy")

    np.testing.assert_array_almost_equal(cropped_grain_image, expected, decimal=4)
