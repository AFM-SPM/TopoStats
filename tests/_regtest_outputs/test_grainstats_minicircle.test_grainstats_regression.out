                     centre_x      centre_y    radius_min    radius_max   radius_mean  radius_median    height_min    height_max  height_median   height_mean        volume          area  area_cartesian_bbox  smallest_bounding_width  smallest_bounding_length  smallest_bounding_area  aspect_ratio threshold     max_feret     min_feret image
molecule_number                                                                                                                                                                                                                                                                                                                                    
0                7.500895e-08  4.775362e-08  4.044948e-09  2.542964e-08  1.600048e-08   1.647888e-08  1.006306e-09  2.696640e-09   1.597337e-09  1.596062e-09  1.084371e-24  6.794043e-16         1.319762e-15             2.053897e-08              5.037861e-08            1.034725e-15      2.452830     above  5.037861e-08  2.053897e-08  None
1                8.028649e-08  7.895237e-08  6.658610e-09  2.623645e-08  1.609777e-08   1.573693e-08  1.009513e-09  2.480685e-09   1.668932e-09  1.683197e-09  1.031843e-24  6.130257e-16         1.526707e-15             2.017441e-08              4.942135e-08            9.970467e-16      2.449704     above  4.987228e-08  2.017441e-08  None
2                4.001424e-08  7.585689e-08  9.745734e-09  2.369314e-08  1.746942e-08   1.818722e-08  1.006681e-09  2.139365e-09   1.646541e-09  1.598842e-09  1.123717e-24  7.028320e-16         1.546230e-15             3.359220e-08              4.149625e-08            1.393950e-15      1.235294     above  4.440534e-08  3.176819e-08  None
